### PR TITLE
CCv0 | agent: Add PullImage endpoint

### PR DESCRIPTION
**Pull request by stevenhorsman**
This commit adds the PullImge endpoint to the agent
and the agent-ctl command to test it.

Fixes: #2509

Signed-off-by: Georgina Kinge <georgina.kinge@ibm.com>

### CODE REVIEW

**Comment by fidencio** [comment]
I've made the "move-linked-issues-to-in-progress" not required as I'm not sure what it was not working.  I may need to do the same for the tests repo branch.

Now all the CIs are green, good job @stevenhorsman!

File: src/agent/src/rpc.rs [hunk 1]
```diff
@@ -1722,6 +1747,150 @@ fn load_kernel_module(module: &protocols::agent::KernelModule) -> Result<()> {
     }
 }
 
+fn pull_image_from_registry(image: &str, src_creds: &str) -> Result<()> {
+    let source_image = format!("{}{}", "docker://", image);
+
+    // TODO LATER derive directory from image registry/namespace/repository/tag to make this more dynamic
+    // Define the target tranport and path for the manifest image, with signature
```
Comment (sameo) [inline src/agent/src/rpc.rs:1754]: Couldn't we just prefix it with the container ID instead, and pave the way for a full pod images pull?

```Rust
let target_path_manifest = format!("dir://{}/image_manifest", cid);
```

File: src/agent/src/rpc.rs [hunk 2]
```diff
@@ -1722,6 +1747,150 @@ fn load_kernel_module(module: &protocols::agent::KernelModule) -> Result<()> {
...
+    // The image with a signature can then be unpacked into a bundle
+    let status = Command::new(SKOPEO_PATH)
+        .arg("copy")
+        .arg(&target_path_manifest)
+        .arg(&target_path_oci)
+        .arg("--remove-signatures")
+        .status()
+        .expect("Failed to copy image");
+
+    assert!(status.success());
```
Comment (sameo) [inline src/agent/src/rpc.rs:1810]: I'd rather see us properly propagating the `status()` call `Result` through, with a map_err and anyhow! for example. The error from `status` is an `ExitStatus` that implements `Display` fwiw.

File: src/agent/src/rpc.rs [hunk 3]
```diff
@@ -1722,6 +1747,150 @@ fn load_kernel_module(module: &protocols::agent::KernelModule) -> Result<()> {
...
+        .expect("Failed to verify signature");
+
+    assert!(status.success());
+
+    Ok(())
+}
+
+fn unpack_image(cid: &str) -> Result<()> {
+    // To save space delete the manifest. This was only needed as long as verify image
+    let status = Command::new("rm")
```
Comment (sameo) [inline src/agent/src/rpc.rs:1856]: Here as well, we should use https://doc.rust-lang.org/std/fs/fn.remove_dir_all.html instead.

File: src/agent/src/rpc.rs [hunk 4]
```diff
@@ -1722,6 +1747,150 @@ fn load_kernel_module(module: &protocols::agent::KernelModule) -> Result<()> {
...
+
+    assert!(status.success());
+
+    //To save space delete the oci image after unpack
+    let status = Command::new("rm")
+        .arg("-rf")
+        .arg(&IMAGE_OCI)
+        .status()
+        .expect("Failed to delete oci image");
+
```
Comment (sameo) [inline src/agent/src/rpc.rs:1888]: Ditto.

File: versions.yaml [hunk 5]
```diff
@@ -190,6 +190,8 @@ externals:
     # instead of abstract socket, thus kata wouldn's support the containerd's
     # version older than them.
     version: "v1.5.2"
+    # CCv0 is using a PR containerd build as our changes are still PoC
+    pr_id: "5911"
```
Comment (sameo) [inline versions.yaml:194]: Could we push the `tools` and `versions.yaml` changes at least as a separate commit, please :)

**Review by sameo — COMMENTED** [review]
Thanks a lot for the PR @stevenhorsman ! A few code specific comments and in general I think we could aim at something simpler, as I feel the creds and PGP verification might be something that will be overridden later on.

File: versions.yaml (hunk 5 above)
Comment (stevenhorsman) [reply versions.yaml]: I've created PR #2577 for this change

File: src/agent/src/rpc.rs [hunk 6]
```diff
@@ -78,6 +78,10 @@ use std::path::PathBuf;
 
 const CONTAINER_BASE: &str = "/run/kata-containers";
 const MODPROBE_PATH: &str = "/sbin/modprobe";
+const SKOPEO_PATH: &str = "/usr/bin/skopeo";
+const UMOCI_PATH: &str = "/usr/local/bin/umoci";
+#[allow(dead_code)]
+const GPG_HOME: &str = "/tmp/gpg_home";
```
Comment (sameo) [inline src/agent/src/rpc.rs:84]: I think this one is no longer needed.

File: src/agent/src/rpc.rs [hunk 7]
```diff
@@ -1722,6 +1741,78 @@ fn load_kernel_module(module: &protocols::agent::KernelModule) -> Result<()> {
...
+
+    let manifest_path = format!("/tmp/{}/image_manifest", cid);
+    let target_path_manifest = format!("dir://{}", manifest_path);
+
+    // Define the target transport and path for the OCI image, without signature
+    let oci_path = format!("/tmp/{}/image_oci:latest", cid);
+    let target_path_oci = format!("oci://{}", manifest_path);
+
+    let _ = fs::create_dir_all(&manifest_path);
+    let _ = fs::create_dir_all(&oci_path);
```
Comment (sameo) [inline src/agent/src/rpc.rs:1755]: I think just `create_dir_all()?;` will work.

File: src/agent/src/rpc.rs [hunk 8]
```diff
@@ -1722,6 +1741,78 @@ fn load_kernel_module(module: &protocols::agent::KernelModule) -> Result<()> {
...
+
+    let status = Command::new(SKOPEO_PATH)
+        .arg("copy")
+        .arg(source_image)
+        .arg(&target_path_manifest)
+        .status()
```
Comment (sameo) [inline src/agent/src/rpc.rs:1761]: Sorry I was wrong in my initial review. What I had in mind was:

```Rust
   let status: ExitStatus = Command::new(SKOPEO_PATH)
        .arg("copy")
        .arg(source_image)
        .arg(&target_path_manifest)
        .status()?;

    if !status.success() {
        return Err(anyhow!(format!("failed to pull image: {:?}", status)));
    }

```

File: tools/osbuilder/rootfs-builder/fedora/config.sh [hunk 9]
```diff
@@ -5,7 +5,7 @@
 
 OS_NAME="Fedora"
 
-OS_VERSION=${OS_VERSION:-30}
+OS_VERSION=${OS_VERSION:-34}
```
Comment (sameo) [inline tools/osbuilder/rootfs-builder/fedora/config.sh:8]: Is this needed to get the umoci and skopeo packages?

Comment (stevenhorsman) [reply tools/osbuilder/rootfs-builder/fedora/config.sh:8]: I think Dave said that The version of skopeo in 30 is very back level and Fedora 30 is end of life: #2116

**Comment by stevenhorsman** [comment]
Thanks for the review comments Samuel and being patient when we display our lack of rust experience. I'm still working on the changes as skopeo doesn't seem to like the new path names, so I'm trying to debug and work through that.

**Comment by stevenhorsman** [comment]
Hey @sameo - after way too long, I think I've implemented all the changes you requested and tracked down all of the problems, so it is ready for a re-review, but as you have spotted my rust isn't great, so please add any more comments as I'm eager to learn. Thanks!

**Review by sameo — APPROVED** [review]
Thanks @stevenhorsman for the quick fixes!

**Pull request merged**